File: tsp.cpp
#include "tsp.h"
#include "tour.h"
#include <iostream>
#include <math.h>
#include <assert.h>
#include <stdio.h>
#define DEBUG 1


/**
 * Constructs a new TSP instance
 */
TSP::TSP(int nodes) : numNodes(nodes) {
    tour = new Tour();
}

/**
 * Destructs this TSP instance.
 */
TSP::~TSP() {
    delete tour;
}


/*
 * Läser in alla noder och lägger dem i nodes[]
 */
void TSP::init() {
    float x,y;
    for (int i = 0; i < numNodes; i++) {
        scanf("%f", &x);
        scanf("%f", &y);
        nodes.push_back(new node(x,y));
        neighbors.push_back(new std::multimap<float, short int>);
    }
}

/*
 * Skapar alla grannlistor
 */
void TSP::createNeighbors() {
    // Vi letar reda på alla grannar vi har
    for (int i = 0; i < numNodes; i++) {
        for (int n = 0; n < numNodes; n++) {
            if (i != n) {
                neighbors[i]->insert(std::pair<float, int>(
                            distance(i, n),
                            n));
                // om vi har fler grannar än vi är intresserade av 
                // tar vi bort den sista
                if (neighbors[i]->size() > neighborsToCheck) {
                    neighbors[i]->erase(--neighbors[i]->end());
                }
            }
        }
    }
}

/*
 * Räknar ut avståndet mellan de två noderna a och b
 */
float TSP::distance(const int a, const int b) const {
    return sqrt(pow((nodes[b]->x - nodes[a]->x), 2) + pow((nodes[b]->y - nodes[a]->y),2));
}



/*
 * Håll koll på vart i en tur en nod är
 *
 * Vi behöver detta för att kunna slå upp saker i O(1) tid 
 * när vi ska titta på grannar, iom att vi inte vet vart i turen
 * en granne finns
 */
void TSP::createPos() {
    for (int i = 0; i < numNodes; i++) {
        (*pos)[(*tour)[i]] = i;
    }
}

/*
 * Printar ut alla grannar
 *
 */
void TSP::printNeighbors() {
    std::multimap<float, short int>::iterator iter;

    for (int i = 0; i < numNodes; i++) {
        std::cerr << i << " { "<< std::endl;
        for (iter = neighbors[i]->begin(); iter != neighbors[i]->end(); iter++) {
            std::cerr << "\t" << iter->first << " => " << iter->second << std::endl;
        }
        std::cerr << i << " } " << std::endl;
    }
}

void TSP::greedyPath() {
    bool used[numNodes];
    // initialize used to false
    // and tour to 0;
    for (int i = 0; i < numNodes; i++) {
        used[i] = false;
        tour->add(0); 
    }

    (*tour)[0] = 0;
    used[0] = true;
    int best;

    for (int i = 1; i < numNodes; i++) {
        best = -1;
        for (int j = 0; j < numNodes; j++) {
            if (!used[j]) {
                if (best == -1 || (distance((*tour)[i-1], j) < distance((*tour)[i-1], best))) {
                    best = j;
                }
            }
        }
        (*tour)[i] = best;
        used[best] = true;
    }
}

void TSP::oneToN() {
    for (int i = 0; i < numNodes; i++) {
        tour->add(i);
    }
}

/**
 * Performs the 2-opt local search.
 */
void TSP::twoOpt() {

}

void TSP::printTour() {
    std::cout << (*tour);
}

float TSP::tourLength() {
    return tour->length(*this);
}

int main() {
    int nodes;
    scanf("%d", &nodes);
    if (DEBUG)
        std::cerr << "numNodes:\t" << nodes << std::endl;
    TSP tsp(nodes);

    tsp.init();
    std::cerr << "init done" << std::endl;
    tsp.createNeighbors();
    std::cerr << "createNeighbors() done" << std::endl;
    // tsp.printNeighbors();
    // std::cerr << "printNeighbors() done" << std::endl;

    // dont forget to init tour somewhere

    
//    tsp.oneToN();

    std::cerr << "oneToN() done" << std::endl;
    tsp.greedyPath();

    std::cerr << "greedyPath() done" << std::endl;

    // perform local search optimization
    tsp.twoOpt();
    std::cerr << "optimization done" << std::endl;

    tsp.printTour();

    std::cerr << "printTour() done" << std::endl;

    std::cerr << "tour length:\t" << tsp.tourLength() << std::endl;

}

File: TSP.java
import java.lang.Math;
import java.util.Arrays;
public class TSP {
    Kattio io;
    int numNodes;
    Node[] nodes;
    int[][] neighbors;

    double[][] distance;

    Tour tour;

    public final int DEBUG = 0;

    private int neighborsToCheck = 5;

    public static void main (String[] argv) {
        new TSP();
    }
    TSP() {
        io = new Kattio(System.in);
        this.numNodes = io.getInt();
        nodes = new Node[numNodes];
        tour = new Tour(numNodes);
        long time;
        if (DEBUG > 0)
            time = System.currentTimeMillis();

        readNodes();

        if (DEBUG > 0) {
            System.err.println("readNodes()         " + (System.currentTimeMillis() - time));	
            time = System.currentTimeMillis();
        }

        createDistance();
        if (DEBUG > 0) {
            System.err.println("createDistance()	" + (System.currentTimeMillis() - time));
            time = System.currentTimeMillis();
        }

        createNeighbors();  
        if (DEBUG > 0) {
            System.err.println("createNeighbors()	" + (System.currentTimeMillis() - time));
            time = System.currentTimeMillis();
        }

        // printNeighbors();

        NNPath();
        if (DEBUG > 0) {
            System.err.println("NNPath()	        " + (System.currentTimeMillis() - time));
            time = System.currentTimeMillis();
        }

        twoOpt();
        if (DEBUG > 0) {
            System.err.println("twoOpt()	        " + (System.currentTimeMillis() - time));
            time = System.currentTimeMillis();
        }

        twoDotFiveOpt();
        if (DEBUG > 0) {
            System.err.println("twoDotFiveOpt()	        " + (System.currentTimeMillis() - time));
            time = System.currentTimeMillis();
        }

        printTour();
        if (DEBUG > 0) {
            System.err.println("printTour()	        " + (System.currentTimeMillis() - time));
            System.err.println("tour length:  " + tour.length(this));
        }

        io.close();
    }

    void readNodes() {
        for (int i = 0; i < numNodes; i++) {
            nodes[i] = new Node(io.getDouble(), io.getDouble());
        }
    }

    void createDistance() {
        distance = new double[numNodes][numNodes];
        for (int i = 0; i < numNodes; i++) {
            for (int j = i; j < numNodes; j++) {
                distance[i][j] = calcDistance(i,j);
                distance[j][i] = distance[i][j];
            }
        }
    }

    /**
     * Försöker skapa grannar
     */
    void createNeighbors() {
        if (neighborsToCheck >= numNodes)
            neighborsToCheck = numNodes-1;
        neighbors =  new int[numNodes][neighborsToCheck];

        // int tmp, prev = 0;

        double dist;
        for (int i = 0; i < numNodes; i++) {
            // Möjlig optimering, sätt n = i 
            Arrays.fill(neighbors[i], -1);
innerFor:
            for (int n = 0; n < numNodes; n++) {
                if (i != n) {
                    dist = distance(i, n);

                    // om den är en värdig granne, sätt in den 
                    if (neighbors[i][neighborsToCheck-1] != -1 && dist > distance(i, neighbors[i][neighborsToCheck-1]))
                        continue;

                    
                    int min=0, max, mid;
                    max = neighborsToCheck;

                    while(max-min <= 4) {
                        mid = (min + max)/2;
                        if (dist > distance(i, neighbors[i][mid])) 
                            min = mid +1;
                        else
                            max = mid -1;
                    }

                    // Vi kollar om noden är en värdig granne
                    for (int j = min; j < max; j++) {

                        // Är detta en bättre granne?
                        if (neighbors[i][j] == -1 || (distance(i, neighbors[i][j]) > dist)) {

                            int k = neighborsToCheck-1;
                            while (k > j) {
                                neighbors[i][k] = neighbors[i][k-1];
                                k--;
                            }
                            neighbors[i][j] = n;
                            continue innerFor;

                        }
                    }
                }
            }
        }
    }

    /**
     * Vi kanske kan tjäna lite hastighet här genom att buffra outputen
     * Såhär?
     */
    void printTour() {
        io.println(tour);
        return;
    }

    /**
     * Print all neighbors in a matrix style kinda way
     */
    void printNeighbors() {
        if (DEBUG <= 0) return;

        System.err.println("=== Neighbors ===");
        for (int i = 0; i < numNodes; i++) {
            for (int j = 0; j < neighborsToCheck; j++) {
                // System.err.printf("%d \t", neighbors[i][j]);
                System.err.printf("%d (%f)\t", neighbors[i][j], distance(i, neighbors[i][j]));
            }
            System.err.println();
        }
    }

    /* eventuell optimering, ´cacha alla avstånt i en matris 
     * (sätt okalkylerade värden till -1)
     */
    double calcDistance(int a, int b) {
        return Math.sqrt((nodes[a].x-nodes[b].x) * (nodes[a].x-nodes[b].x) + (nodes[a].y-nodes[b].y) * (nodes[a].y-nodes[b].y));
    }

    double distance(int a, int b) {
        return distance[a][b];
    }

    void NNPath() {
        boolean used[] = new boolean[numNodes];
        int best;

        tour.addNode(0);
        used[0] = true;

        for (int i = 1; i < numNodes; i++) {
            best = -1;
            for (int j = 0; j < numNodes; j++) {
                if (!used[j] && (best == -1 || 
                            distance(tour.getNode(i-1), j) < distance(tour.getNode(i-1), best))) {
                    best = j;
                            }
            }
            tour.addNode(best); // should be on pos i
            used[best] = true;
        }
    }

    void twoOpt() {
        boolean improvement = true;
        while (improvement) {
            improvement = false;
improve: // restart 
            for (int i = 0; i < numNodes; i++) {
                int c1 = tour.getNode(i); // citi 1
                int nc1 = tour.getNextNode(i); // the next city after city 1
                
                // select next edge from neighbor list
                for (int n = 0; n < neighborsToCheck; n++) {
                    int c2 = neighbors[c1][n];
                    int nc2 = tour.getNextNode(tour.getPos(c2));

                    if (c1 == c2) // doesn't happen if neighbors[][] isn't borken
                        continue;
                    
                    // printTour();

                    // TODO: bra?
                     if (distance(c1, c2) < distance(c1, nc1) ||
                        distance(c1, nc1) < distance(c2, nc2)) {
                        if (distance(c1, nc1) + distance(c2, nc2) >
                            distance(c1, c2) + distance(nc1, nc2)) {
                            improvement = true;
                            reverse(tour.getPos(nc1), tour.getPos(c2));
                            continue improve;
                        }
                    }
                }
            }
        }
    }

    void twoDotFiveOpt() {
        boolean improvement = true;
        while (improvement) {
            improvement = false;
improve: // restart 
            for (int i = 0; i < numNodes; i++) {
                int t1 = tour.getNode(i); // city 1
                int t2 = tour.getNextNode(i); // follows city 1 immediately 
                
                // find next city (t3) from neighbor list (and thereby t4)
                for (int n = 0; n < neighborsToCheck; n++) {
                    int t3 = neighbors[t1][n];
                    
                    // throw away seemingly useless moves
                    if (t1 == t3) continue;
                    if (t2 == t3) continue; 

                    int t4 = tour.getNodeAfter(t3);
                    int t5 = tour.getNodeAfter(t2); // rätt tänkt?

                    if (t1 == t4) continue;
                    if (t2 == t4) continue;
                    if (t3 == t5) continue;
                    
                    // 2.5-opt
                    double curr = distance(t1,t2) + distance(t3,t4) + distance(t2,t5);
                    double opt25 = distance(t4,t2) + distance(t3,t2) + distance(t1,t5);

                    if (opt25 < curr) {
                        tour.moveBetween(t2, t3, t4);
                        improvement = true;
                        continue improve;
                    } /*else if (distance(t1, t3) < distance(t1, t2) ||
                        distance(t1, t2) < distance(t3, t4)) {
                        if (distance(t1, t2) + distance(t3, t4) >
                            distance(t1, t3) + distance(t2, t4)) {
                            improvement = true;
                            reverse(tour.getPos(t2), tour.getPos(t3));
                            continue improve;
                        }
                    }
                    */
                    
                    /*else if ( distance(t1, t2) + distance(t3, t4) >
                                distance(t1, t3) + distance(t2, t4) )
                    {
                        reverse(tour.getPos(t2), tour.getPos(t3));
                        improvement = true;
                        continue improve;
                    }
                    blir sämre utan
                    */
                }
            }
        }
    }

    public void reverse(int start, int end)
    {
        //System.err.println("reverse\t"+start+"\t"+end);
        int len = end - start;

        if (len < 0) // vi har en wraparound
        {
            len += numNodes;
        }

        len = (len+1)/2; // vi behöver bara swappa "hälften av längden" ggr, eftersom vi tar 2 element per swap (+1 för att inte udda antal element ska lämnas kvar)

        for (int k = 0; k < len; k++)
        {
            tour.swap(start, end);

            // wraparound för start & end
            if (++start == numNodes)
                start = 0;

            if (--end <= 0)
                end = numNodes-1;
        }
    }

}
